src/zip_code_tree.hpp

```diff
@@ -203,110 +325,145 @@ class ZipCodeTree {
-        inline bool operator!=(const iterator& other) const {
-            return !(*this == other);
-        }
+        inline bool operator!=(const seed_iterator& other) const { return !(*this == other); }
+
+        /// What item does index point to?
+        tree_item_t current_item() const { return zip_code_tree.at(index); }
         
         /// Get the index and orientation of the seed we are currently at.
-        oriented_seed_t operator*() const;
+        /// Also return all other seeds on the same position
+        vector<oriented_seed_t> operator*() const { return current_seeds; }
 
-        /// Get the number of tree storage slots left in the iterator. We need
-        /// this to make reverse iterators from forward ones.
-        size_t remaining_tree() const;
+        /// Getters
+        /// We need these to make reverse iterators from forward ones.
+        inline size_t get_index() const { return index; }
+        inline bool get_right_to_left() const { return right_to_left; }
+        inline stack<size_t> get_chain_numbers() const { return chain_numbers; }
 
     private:
+        /// Within each parent snarl, which chain are we in?
+        stack<size_t> chain_numbers;
         /// Where we are in the stored tree.
-        vector<tree_item_t>::const_iterator it;
-        /// Where the stored tree ends. We keep this to avoid needing a reference back to the ZipCodeTree.
-        vector<tree_item_t>::const_iterator end;
+        size_t index;
+        /// How far inside of a cyclic snarl are we?
+        size_t cyclic_snarl_nestedness;
+        /// Direction a distance iterator should be set off
+        /// NOT the direction this iterator moves;
+        /// a seed_iterator always moves left to right
+        bool right_to_left;
+        /// Seeds at this position
+        vector<oriented_seed_t> current_seeds;
     };
 
     /// Get an iterator over indexes of seeds in the tree, left to right.
-    iterator begin() const;
+    seed_iterator begin() const { return seed_iterator(0, *this); }
     /// Get the end iterator for seeds in the tree, left to right.
-    iterator end() const;
+    /// (Note that the last element will never be a seed)
+    seed_iterator end() const { return seed_iterator(zip_code_tree.size() - 1, *this); }
 
     /**
-     * Iterator that looks left in the tree from a seed, possibly up to a maximum base distance.
+     * Iterator that looks sideways in the tree from a seed,
+     * possibly up to a maximum base distance.
      *
-     * See https://github.com/benedictpaten/long_read_giraffe_chainer_prototype/blob/b590c34055474b0c901a681a1aa99f1651abb6a4/zip_tree_iterator.py.
+     * Original: 
+     * https://github.com/benedictpaten/long_read_giraffe_chainer_prototype/blob/b590c34055474b0c901a681a1aa99f1651abb6a4/zip_tree_iterator.py.
+     * 
+     * New cyclic snarl handling: 
+     * https://docs.google.com/drawings/d/1diKMXCuMteR06fF64BhSttsD5RHh3eTnd8FtyYqs17Y/edit?usp=sharing
```

Comment:
Is there a writeup of the approach to iterating over cyclic snarls, besides what's in the PR comments? Someone might come here for a general overview of the approach, and while e.g. the diagram is helpful, it doesn't answer questions like "should we really be toggling orientation first thing in `operator++`?".